README.md
# NAME

App::CpanTestReport - Search for CPAN Testers reports by module, distribution or author.

# SYNOPSIS

    $ ./script/cpantestreport daemon --listen http://*:3000

# DESCRIPTION

[App::CpanTestReport](https://metacpan.org/pod/App::CpanTestReport) is a web page where you can search for CPAN Testers
reports by module, distribution or author and see test reports created by
[http://cpantesters.org/](http://cpantesters.org/).

You can see an example of this application at
[https://app.thorsen.pm/cpantestreport](https://app.thorsen.pm/cpantestreport).

# RESOURCES

- /

    Show a search page.

- /search

    Alias for "/".

- /search?q=JHTHORSEN
- /search?author=JHTHORSEN

    Redirects to author search results.

- /search?dist=Mojolicious
- /search?dist=Mojolicious-8.01
- /search?q=Mojolicious
- /search?q=Mojolicious-8.01

    Redirects to dist search results.

- /author/JHTHORSEN

    Show a summary of all the dists for a given author.

- /dist/Mojolicious

    Redirects to the latest version of [Mojolicious](https://metacpan.org/pod/Mojolicious)

- /dist/Mojolicious-8.01

    Show search results for [Mojolicious](https://metacpan.org/pod/Mojolicious) version 8.01.

- /report/57fffae2-689f-1015-8f24-a8e65f496936

    Show a given test report.

# METHODS

## startup

Called by [Mojolicious](https://metacpan.org/pod/Mojolicious) to start up the web server.

# AUTHOR

Jan Henning Thorsen

# COPYRIGHT AND LICENSE

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

# SEE ALSO

[http://api.cpantesters.org/](http://api.cpantesters.org/),
[https://api.metacpan.org/](https://api.metacpan.org/) and
[https://github.com/cpan-testers/cpantesters-web](https://github.com/cpan-testers/cpantesters-web).


Changes
Revision history for perl distribution App-CpanTestReport

0.04 Not Released
 - Add support for X-Request-Base header
 - Fix parsing new email format

0.03 2018-12-19T08:51:12+0900
 - Fix searching for module names with multiple "::"
 - Add tests
 - Improved styling for small screens

0.02 2018-10-05T18:06:47+0900
 - Fix finding correct lib directory
 - Add missing dependency Text::Markdown
 - Replaced PromiseActions with custom hook
 - Add custom 500 and 404 pages

0.01 2018-10-04T16:49:15+0900
 - Add cache and backend helpers
 - Can show summary for author
 - Can redirect to latest dist version
 - Can render test reports for dist
 - Can render individual test report
 - Ready for docker



lib/App/CpanTestReport/Controller/Test.pm
package App::CpanTestReport::Controller::Test;
use Mojo::Base 'Mojolicious::Controller';

sub report {
  my $self = shift;

  return $self->_get_report_p($self->stash('guid'))->then(sub {
    my $report = shift;
    my $text   = delete $report->{result}{output}{uncategorized};
    $self->respond_to(json => {json => $report}, txt => {text => $text}, any => $report);
  });
}

sub _clean {
  local $_ = shift;
  s!^\W+!!;
  s!\W+$!!;
  return $_;
}

sub _get_report_p {
  my ($self, $guid) = @_;
  return $self->backend(cpantesters => [report => $guid])->then(sub { $self->_parse_report(shift) });
}

sub _parse_report {
  my ($self, $report) = @_;
  my @lines   = split /\n\r?/, $report->{result}{output}{uncategorized} || '';
  my $section = 'preamble';
  my $sub_section;

  $report->{result}{environment_and_other_context} //= '';
  $report->{result}{preamble}                      //= '';
  $report->{result}{prerequisites} = {};
  $report->{result}{program_output}  //= '';
  $report->{result}{tester_comments} //= '';

  while (defined(my $line = shift @lines)) {
    if ($line =~ /^Sections of this report/) {
      $report->{result}{$section} ||= _clean($report->{block});
      $report->{block} = '';
      $section = 'toc';
    }
    elsif ($line =~ /^PREREQUISITES/i) {
      $report->{result}{$section} ||= _clean($report->{block});
      $section = 'prerequisites';
    }
    elsif ($line =~ /^TEST RESULTS/i) {
      $report->{result}{$section} ||= _clean($report->{block});
      $section = 'program_output';
      $report->{block} = '';
    }
    elsif ($line =~ /^-----+$/ and $lines[1] eq $line) {
      $report->{result}{$section} ||= _clean($report->{block});
      $report->{block} = '';
      local $_ = shift @lines;
      next unless /(\w.*)/;
      $section = lc $1;
      $section =~ s!\s+!_!g;
    }
    elsif ($section eq 'prerequisites') {
      $sub_section = $1 if $line =~ /^(\w+):/;
      $sub_section ||= 'requires';
      $report->{result}{$section}{$sub_section}{$1} = $2
        if $line =~ m!(\S+)\s+([\d\.]+)\s+([\d\.]+)! or $line =~ m!(\S+)\s+([\d\.]+)!;
    }
    else {
      $report->{block} .= "$line\n";
    }

    # warn "--- $section <<< ($line)\n";
  }

  #warn Mojo::Util::dumper($report);

  $report->{reporter}{email} ||= '';
  $report->{reporter}{email}         =~ s!\s(.*)!!;
  $report->{result}{tester_comments} =~ s!Additional comments from \S+!!;
  $report->{result}{$section} ||= _clean($report->{block});    # Capture last block
  delete $report->{block};
  return $report;
}

1;

=encoding utf8

=head1 NAME

App::CpanTestReport::Controller::Test - Test actions

=head1 METHODS

=head2 report

Will show a given test report.

=head1 SEE ALSO

L<App::CpanTestReport>.

=cut


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'App::CpanTestReport',
  AUTHOR         => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/App/CpanTestReport.pm',
  VERSION_FROM   => 'lib/App/CpanTestReport.pm',
  EXE_FILES      => [qw(script/cpantestreport)],
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {'Test::More' => '0.88'},
  PREREQ_PM      => {'Mojo::Redis' => '3.11', 'Mojolicious' => '8.00'},
  META_MERGE     => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/jhthorsen/app-cpantestreport/issues'},
      homepage   => 'https://github.com/jhthorsen/app-cpantestreport',
      repository => {
        type => 'git',
        url  => 'https://github.com/jhthorsen/app-cpantestreport.git',
        web  => 'https://github.com/jhthorsen/app-cpantestreport',
      },
    },
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/App/CpanTestReport/Controller/Dist.pm
package App::CpanTestReport::Controller::Dist;
use Mojo::Base 'Mojolicious::Controller';

my $VERSION_RE = qr{[\d\.]+};

sub report {
  my $self = shift;

  my ($name, $version, $format) = $self->stash('name') =~ m!^(.+)-($VERSION_RE)(?:\.(html|json))?$!;
  ($name, $format) = $self->stash('name') =~ m!^(.+)(?:\.(html|json))?$! unless $name;
  $self->stash(format => $format) if $format;

  if (!$name) {
    return $self->reply->not_found;
  }
  elsif ($name and $version) {
    $self->param(q => "$name-$version");
    return $self->_get_report_p({name => $name, version => $version}, $self->req->url->query->to_hash)->then(sub {
      my $res = shift;
      $self->respond_to(json => {json => $res}, any => $res);
    });
  }
  else {
    return $self->_get_latest_dist_p($name)->then(sub {
      return $self->reply->not_found unless my $dist = shift;
      return $self->redirect_to($self->url_for('dist.report', {format => $self->stash('format'), name => $dist}));
    });
  }
}

sub _get_latest_dist_p {
  my ($self, $module) = @_;

  $module =~ s!-!::!g;    # Turn Dist-Name into Module::Name
  return $self->backend(metacpan => [module => $module])->then(sub {
    my $module = shift;
    return undef unless $module->{distribution} and $module->{version} and $module->{version} =~ /$VERSION_RE$/;
    return join '-', @$module{qw(distribution version)};
  });
}

sub _get_report_p {
  my ($self, $args, $filters) = @_;

  return $self->backend(cpantesters => [summary => @$args{qw(name version)}])->then(sub {
    my $reports = shift;

    if (ref $reports ne 'ARRAY' or !@$reports) {
      my $error = ref $reports eq 'HASH' ? $reports->{errors}[0]{message} : 'No reports found.';
      $error .= '.' unless $error =~ /[^a-z]$/i;
      return {error => $error, reports => [], stats => {}};
    }

    my (@filtered, %stats);
    for my $report (@$reports) {
      my @perl_v = map { /^\d+$/ ? int : $_ } split /\./, $report->{perl};
      my $perl_m = join '.', @perl_v[0, 1];
      $report->{platform} =~ s!^(?:\w+\.)?(\w+)-.*$!$1!;
      $stats{grade}{$report->{grade}}++;
      $stats{osname}{$report->{osname}}{$report->{grade}}++;
      $stats{perl}{$perl_m}{$report->{grade}}++;
      $stats{platform}{$report->{platform}}{$report->{grade}}++;

      next unless _in($report->{grade},    $filters->{grade});
      next unless _in($report->{osname},   $filters->{osname});
      next unless _in($report->{platform}, $filters->{platform});
      next unless _in($perl_m, $filters->{perl}) or _in($report->{perl}, $filters->{perl});

      delete @$report{qw(dist reporter version)};
      $report->{perl} = \@perl_v;
      push @filtered, $report;
    }

    @filtered = sort {
           $a->{grade} cmp $b->{grade}
        || $a->{osname} cmp $b->{osname}
        || $a->{osvers} cmp $b->{osvers}
        || $b->{perl}[1] <=> $a->{perl}[1]
    } @filtered;

    return {reports => \@filtered, stats => \%stats};
  });
}

sub _in {
  my ($needle, $filter) = (shift // '', shift);
  return 1 unless defined $filter;
  return int grep { $needle eq $_ } ref $filter ? @$filter : ($filter);
}

1;

=encoding utf8

=head1 NAME

App::CpanTestReport::Controller::Dist - Distribution actions

=head1 METHODS

=head2 report

Show a report for a distribution with version or redirect to the latest version
of the distribution if version is missing.

=head1 SEE ALSO

L<App::CpanTestReport>.

=cut


lib/App/CpanTestReport/Controller/Root.pm
package App::CpanTestReport::Controller::Root;
use Mojo::Base 'Mojolicious::Controller';

use Mojo::Util 'trim';

sub search {
  my $self  = shift;
  my $query = $self->_normalize_search_query;
  my %attrs = (format => $self->stash('format'));

  if ($query->{author}) {
    return $self->redirect_to('author.summary', %attrs, id => $query->{author});
  }
  elsif ($query->{dist} and $query->{version}) {
    return $self->redirect_to('dist.report', %attrs, name => join '-', @$query{qw(dist version)});
  }
  elsif ($query->{dist}) {
    return $self->redirect_to('dist.report', %attrs, name => $query->{dist});
  }

  my @examples = (
    $self->url_for('/search.json')->query([q      => 'Mojolicious'])->to_abs,
    $self->url_for('/search.json')->query([q      => 'Mojolicious 8.01'])->to_abs,
    $self->url_for('/search.json')->query([q      => 'Mojolicious-8.01'])->to_abs,
    $self->url_for('/search.json')->query([q      => 'JHTHORSEN'])->to_abs,
    $self->url_for('/search.json')->query([author => 'JHTHORSEN'])->to_abs,
    $self->url_for('/search.json')->query([dist   => 'Mojolicious'])->to_abs,
    $self->url_for('/search.json')->query([dist   => 'Mojolicious', version => '8.01'])->to_abs,
    $self->url_for('author.summary', %attrs, id   => 'JHTHORSEN')->to_abs,
    $self->url_for('dist.report',    %attrs, name => 'Mojolicious-8.01')->to_abs,
    $self->url_for('dist.report',    %attrs, name => 'Mojolicious-8.01')
      ->query([osname => 'linux', grade => 'unknown', grade => 'na', platform => 'x86_64', perl => '5.6', perl => '5.24'
      ])->to_abs,
  );

  $self->respond_to(json => {json => {examples => \@examples}}, any => {examples => \@examples});
}

sub _normalize_search_query {
  my $c     = shift;
  my $query = $c->req->url->query->to_hash;

  for my $p (grep {/\w/} map { trim $_ } split /\s+/, delete $query->{q} // '') {
    if ($p =~ /^(.+)-([\d\.]+)/) {
      $query->{dist}    ||= $1;
      $query->{version} ||= $2;
    }
    if ($p =~ /^\d/) {
      $query->{version} ||= $p;
    }
    elsif ($p =~ /^[A-Z]+$/) {
      $query->{author} ||= $p;
    }
    else {
      $query->{dist} ||= $p;
      $query->{dist} =~ s!::!-!g;
    }
  }

  return $query;
}

1;

=encoding utf8

=head1 NAME

App::CpanTestReport::Controller::Root - Various actions

=head1 METHODS

=head2 search

Will show a search page or redirect to a given resource based on input.

=head1 SEE ALSO

L<App::CpanTestReport>.

=cut


lib/App/CpanTestReport/Controller/Author.pm
package App::CpanTestReport::Controller::Author;
use Mojo::Base 'Mojolicious::Controller';

use List::Util 'sum';

sub summary {
  my $self = shift->render_later;

  return $self->_get_summary_by_author_p($self->stash('id'))->then(sub {
    my $res = shift;
    $self->param(q => $self->stash('id'));
    $self->respond_to(json => {json => $res}, any => {page => 'author', %$res});
  });
}

sub _get_summary_by_author_p {
  my ($self, $author) = @_;

  my @metacpan = (
    metacpan => [qw(release _search)],
    [fields => 'distribution,version', size => '100', q => sprintf('author:%s AND status:latest', uc $author)],
  );

  my %latest;
  return $self->backend(@metacpan)->then(sub {
    my $hits = shift;
    $hits = $hits->{hits} while ref $hits eq 'HASH' and $hits->{hits};
    $latest{$_->{fields}{distribution}} = $_->{fields}{version} for @$hits;

    return $self->backend(cpantesters => [release => author => uc $author]);
  })->then(sub {
    my $json = shift;
    my (@dists, %stats);

    for my $r (sort { $b->{fail} <=> $a->{fail} || $a->{dist} cmp $b->{dist} } @$json) {
      next unless $latest{$r->{dist}} and $r->{version} eq $latest{$r->{dist}};
      $r->{name} = delete $r->{dist};
      $stats{$_} += $r->{$_} || 0 for qw(fail na pass unknown);
      push @dists, $r;
    }

    return {dists => \@dists, stats => \%stats};
  });
}

1;

=encoding utf8

=head1 NAME

App::CpanTestReport::Controller::Author - Author actions

=head1 METHODS

=head2 summary

Show summary for an author.

=head1 SEE ALSO

L<App::CpanTestReport>

=cut


lib/App/CpanTestReport.pm
package App::CpanTestReport;
use Mojo::Base 'Mojolicious';

use Mojo::Redis;
use Scalar::Util 'blessed';
use Time::Piece;
use Time::Seconds;

our $VERSION = '0.03';

sub startup {
  my $self = shift;

  $self->defaults(
    backends => {cpantesters => 'http://api.cpantesters.org/v3', metacpan => 'https://fastapi.metacpan.org/v1'},
    error    => ''
  );

  $self->hook(around_action => \&_hook_around_action);
  $self->hook(before_dispatch => \&_hook_before_dispatch) if $ENV{X_REQUEST_BASE};
  $self->_add_helper_backend;
  $self->_add_helper_cache;
  $self->_add_helper_dist;
  $self->_add_helper_human_date;
  $self->_add_helper_markdown;

  my $r = $self->routes;
  $r->get('/')->to('root#search', page => 'search')->name('search');
  $r->get('/search')->to('root#search', page => 'search');
  $r->get('/author/:id')->to('author#summary', page => 'author')->name('author.summary');
  $r->get('/dist/*name')->to('dist#report',    page => 'dist')->name('dist.report');
  $r->get('/report/:guid')->to('test#report', page => 'report')->name('test.report');
}

sub _add_helper_backend {
  my $self = shift;
  my $ua = $ENV{MOJO_REDIS_CACHE_OFFLINE} ? Mojo::UserAgent->with_roles('+Cache')->new : $self->ua;

  $self->helper(
    backend => sub {
      my ($c, $backend, $path, $query) = @_;

      my $url = Mojo::URL->new($c->stash->{backends}{$backend});
      push @{$url->path}, @$path;
      $url->query($query) if $query;

      return $c->cache->compute_p(
        $url->clone->scheme(undef)->to_string,
        $c->stash('cache_timeout') || 600,
        sub {
          return $ua->get_p($url)->then(sub {
            my $tx = shift;
            return $tx->res->json unless my $err = $tx->error;
            die "[$backend] $err->{message} <<< $url";
          });
        }
      );
    }
  );
}

sub _add_helper_cache {
  my $self  = shift;
  my $redis = Mojo::Redis->new;

  $self->helper(
    cache => sub {
      my $c = shift;
      return $c->stash->{'redis.cache'}
        ||= $redis->cache(namespace => 'cpantestreport')->refresh($c->param('_refresh'));
    }
  );
}

sub _add_helper_dist {
  my $self = shift;
  $self->helper(
    dist => sub {
      local $_ = $_[1];
      s!::!-!g;
      s!-0+$!!;
      return $_;
    }
  );
}

sub _add_helper_human_date {
  my $self = shift;

  $self->helper(
    'human.date' => sub {
      my $date = Mojo::Date->new(pop);
      my $now  = time;

      return localtime($date->epoch)->strftime($date->epoch < $now - ONE_DAY * 180 ? '%e. %b %Y' : '%e. %b, %H:%M');
    }
  );
}

sub _add_helper_markdown {
  my $self = shift;

  $self->helper(
    markdown => sub {
      my ($c, $text) = @_;
      my $dist_url = $c->url_for('dist.report', name => '');

      $text //= '';
      $text =~ s|(\w\S+\@\S+\.\w{2,})|<a href="mailto:$1">$1</a>|g;
      $text =~ s|(https?://\S+)|<a href="$1">$1</a>|g;
      $text =~ s|(\w\S+-\d+\.[\d\.]+)|<a href="$dist_url$1">$1</a>|g;
      $text = join '', map {"<p>$_</p>"} split /\n\r?\n\r?/, $text;

      return Mojo::ByteStream->new($text);
    }
  );
}

sub _hook_around_action {
  my ($next, $c) = @_;
  my @args = $next->();

  $c->inactivity_timeout(20);

  if (blessed($args[0]) and $args[0]->can('then')) {
    my $tx = $c->render_later->tx;
    $args[0]->then(
      undef,
      sub {
        my $err = shift;
        my $reply_with = $err =~ /Not Found/i ? 'not_found' : 'exception';
        $c->reply->$reply_with($err);
        undef $tx;
      }
    );
    $args[0]->wait if $args[0]->can('wait');
  }

  return @args;
}

sub _hook_before_dispatch {
  my $c = shift;
  return unless my $base = $c->req->headers->header('X-Request-Base');
  $c->req->url->base(Mojo::URL->new($base));
}

1;

=encoding utf8

=head1 NAME

App::CpanTestReport - Search for CPAN Testers reports by module, distribution or author.

=head1 SYNOPSIS

  $ ./script/cpantestreport daemon --listen http://*:3000

=head1 DESCRIPTION

L<App::CpanTestReport> is a web page where you can search for CPAN Testers
reports by module, distribution or author and see test reports created by
L<http://cpantesters.org/>.

You can see an example of this application at
L<https://app.thorsen.pm/cpantestreport>.

=head1 RESOURCES

=over 2

=item * /

Show a search page.

=item * /search

Alias for "/".

=item * /search?q=JHTHORSEN

=item * /search?author=JHTHORSEN

Redirects to author search results.

=item * /search?dist=Mojolicious

=item * /search?dist=Mojolicious-8.01

=item * /search?q=Mojolicious

=item * /search?q=Mojolicious-8.01

Redirects to dist search results.

=item * /author/JHTHORSEN

Show a summary of all the dists for a given author.

=item * /dist/Mojolicious

Redirects to the latest version of L<Mojolicious>

=item * /dist/Mojolicious-8.01

Show search results for L<Mojolicious> version 8.01.

=item * /report/57fffae2-689f-1015-8f24-a8e65f496936

Show a given test report.

=back

=head1 METHODS

=head2 startup

Called by L<Mojolicious> to start up the web server.

=head1 AUTHOR

Jan Henning Thorsen

=head1 COPYRIGHT AND LICENSE

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<http://api.cpantesters.org/>,
L<https://api.metacpan.org/> and
L<https://github.com/cpan-testers/cpantesters-web>.

=cut
